### PR TITLE
[libc++][string] Remove potential non-trailing 0-length array

### PR DESCRIPTION
**Pull request by serge-sans-paille**
It is a violation of the standard to use 0 length arrays, especially when not at the end of a structure (not a FAM GNU extension). Compiler generally accept it, but it's probably better to have a conforming implementation.

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
The build error seems unrelated (?)

**Comment by philnik777** [comment]
> It is a violation of the standard to use 0 length arrays, especially when not at the end of a structure (not a FAM GNU extension). Compiler generally accept it, but it's probably better to have a conforming implementation.

It's an extension that compilers currently support, so I don't really see why the new implementation is better. As-is this is a significant additional amount of complexity without much of a downside (other than using a questionable extension, which is purely philosophical).

File: libcxx/include/string
Comment (frederick-vs-ja) [inline libcxx/include/string:762]: `value_type` is not declared here. It should be `_CharT`.
```suggestion
  _CharT __data_[__min_cap];
```

**Review by frederick-vs-ja — CHANGES_REQUESTED** [review]

**Comment by frederick-vs-ja** [comment]
> It's an extension that compilers currently support, so I don't really see why the new implementation is better.

Perhaps the change will make `-Wzero-length-array` happy. #74183 seems related.

**Comment by serge-sans-paille** [comment]
> > It's an extension that compilers currently support, so I don't really see why the new implementation is better.
> 
> Perhaps the change will make `-Wzero-length-array` happy. #74183 seems related.

Exactly the issue I ran into.

**Comment by philnik777** [comment]
> > > It's an extension that compilers currently support, so I don't really see why the new implementation is better.
> > 
> > 
> > Perhaps the change will make `-Wzero-length-array` happy. #74183 seems related.
> 
> Exactly the issue I ran into.

How exactly does one run into a diagnostic in a system header?

**Comment by serge-sans-paille** [comment]
> > > > It's an extension that compilers currently support, so I don't really see why the new implementation is better.
> > > 
> > > 
> > > Perhaps the change will make `-Wzero-length-array` happy. #74183 seems related.
> > 
> > 
> > Exactly the issue I ran into.
> 
> How exactly does one run into a diagnostic in a system header?

I'm working on harmonizing the clang behavior wrt. FMA, and while doing this I turned on FMA warnings when compiling the full llvm-project, looks like it impacted libcxx too.

**Comment by philnik777** [comment]
> I'm working on harmonizing the clang behavior wrt. FMA, and while doing this I turned on FMA warnings when compiling the full llvm-project, looks like it impacted libcxx too.

You're talking about fused-multiply-add, right? Or is there another FMA? What has FMA got to do with anything here?

**Comment by serge-sans-paille** [comment]
> > I'm working on harmonizing the clang behavior wrt. FMA, and while doing this I turned on FMA warnings when compiling the full llvm-project, looks like it impacted libcxx too.
> 
> You're talking about fused-multiply-add, right? Or is there another FMA? What has FMA got to do with anything here?

`s/FMA/FAM/` ^^!

**Comment by serge-sans-paille** [comment]
Gentle ping :-)
There are a few other situations where we use zero length arrays in libc++, and this seem to help for compatibility with msvc as pointed out in #74183

File: libcxx/include/string
Comment (ldionne) [inline libcxx/include/string:769]: Pre-existing but I don't understand it: this struct has no name, is that really intended?

File: libcxx/include/string
Comment (ldionne) [inline libcxx/include/string:753]: I would like to do this:

```c++
// always define both of these classes
template <class _CharT, size_t __min_cap, size_t _Padding = sizeof(_CharT) - 1>
struct __short_layout_alternate { ... };

template <class _CharT, size_t __min_cap, size_t _Padding = sizeof(_CharT) - 1>
struct __short_layout_classic { ... };

// now inside basic_string
#ifdef _LIBCPP_ABI_ALTERNATE_STRING_LAYOUT
using __short = __short_layout_alternate <value_type, __min_cap>;
#else
using __short = __short_layout_classic <value_type, __min_cap>;
#endif
```

This reduces the scope of the conditional code and IMO makes things easier to understand.

**Review by ldionne — COMMENTED** [review]
I have some questions and comments, but this generally LGTM. While I wouldn't go absolutely nuts to avoid using a compiler extension, I think it's better to strive towards using few-ish of those. I also think this is a good opportunity to do a small refactoring of `std::string`'s layout definition that has been on my mind for a while.

@philnik777 Please let me know if you feel strongly that this patch shouldn't move forward. I am weakly favourable to it, but I could also be convinced that the complexity isn't worth it.

File: libcxx/include/string
Comment (philnik777) [reply libcxx/include/string:769]: Yes, it's intended. It's required to keep the ABI on AIX stable.

**Review by philnik777 — COMMENTED** [review]
> I have some questions and comments, but this generally LGTM. While I wouldn't go absolutely nuts to avoid using a compiler extension, I think it's better to strive towards using few-ish of those. I also think this is a good opportunity to do a small refactoring of `std::string`'s layout definition that has been on my mind for a while.
> 
> @philnik777 Please let me know if you feel strongly that this patch shouldn't move forward. I am weakly favourable to it, but I could also be convinced that the complexity isn't worth it.

The main reason I've been pushing back here is because I think there is a significantly simpler alternative and I don't think it's exactly urgent to remove the use of this extension. I know that the clang folks are thinking about removing it (since it is indeed heinous), but I'm not aware of anybody actively working on that at the moment.

Specifically, we could do
```c++
template <size_t _PaddingSize>
struct __padding {
  char __padding_[_PaddingSize];
};

template <>
struct __padding<0> {};
```

and then just have
```c++
struct __short {
  value_type __data_[__min_cap];
  _LIBCPP_NO_UNIQUE_ADDRESS __padding __padding_;
  unsigned char __size_ : 7;
  unsigned char __is_long_ : 1;
};
```
IMO this is significantly more readable.

But it requires the newer AppleClang.

File: libcxx/include/string
Comment (serge-sans-paille) [reply libcxx/include/string]: I've updated the patch accordingly, with a guard on the individual types too.

**Comment by serge-sans-paille** [comment]
gentle ping: I've updated the patch according to @ldionne suggestion. I agree that @philnik777 solution is much cleaner, but requiring AppleClang seems to be a no-go (?)

**Comment by philnik777** [comment]
We've actually update the AppleClang now, so my solution should™ work now.

File: libcxx/include/string
Comment (ldionne) [inline libcxx/include/string:751]: You can get rid of the `#ifdef` here. This type and `__short_layout_classic` can both be defined regardless of which one we end up picking for the actual string representation (based on the single `#ifdef` inside the `basic_string` class).

File: libcxx/include/string
Comment (ldionne) [inline libcxx/include/string:755]: Can you try out @philnik777 's suggestion for the padding? I think it should work with our current CI setup.

File: libcxx/include/string
Comment (philnik777) [inline libcxx/include/string:760]: This shouldn't be required to be a template anymore.

Comment (serge-sans-paille) [reply libcxx/include/string]: Indeed!

**Comment by serge-sans-paille** [comment]
All green again :-)

File: libcxx/include/string
Comment (philnik777) [inline libcxx/include/string:760]: They're still templates.

**Review by ldionne — APPROVED** [review]
I think you might have misunderstood what my refactoring suggestion was, but it doesn't matter. I pushed an update to minimize this diff and I can pursue the refactoring I had in mind separately. LGTM once CI is green.

**Review by philnik777 — APPROVED** [review]

**Comment by serge-sans-paille** [comment]
Thanks @ldionne & @philnik777 for the help on that one :bow:

**Comment by Michael137** [comment]
FYI, looks like this broke couple of tests on the macOS LLDB CI: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/11268/consoleFull
```
********************
  lldb-api :: commands/expression/call-function/TestCallStdStringFunction.py
  lldb-api :: lang/cpp/stl/TestSTL.py


Assertion failed: (M.Offset >= Tail && "Bitfield access unit is not clipped"), function checkBitfieldClipping, file CGRecordLayoutBuilder.cpp, line 960.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	HandleCommand(command = "expression str")
1.	<eof> parser at end of file
2.	<lldb wrapper prefix>:43:1: Generating code for declaration '$__lldb_expr'

-
```

I'm looking into it

**Comment by ldionne** [comment]
I thought we were running the data formatter tests as part of our CI nowadays. Would it behave differently on macOS vs Linux?

**Comment by Michael137** [comment]
> I thought we were running the data formatter tests as part of our CI nowadays. Would it behave differently on macOS vs Linux?

Looks like these two tests are not covered by the data-formatter target, but build and test against libc++. I can audit the test-suite for more of these tests and get them covered by libc++ CI

**Comment by mysterymath** [comment]
This is also breaking Fuchsia CI... could we get a revert or quick fix-forward to get head back to green?

**Comment by Michael137** [comment]
Here's the reduced issue: https://github.com/llvm/llvm-project/pull/108155

I'm working on a fix. But to get CI back in shape, are you ok with reverting this in the meantime @serge-sans-paille? Apologies for the disruption

**Comment by ldionne** [comment]
@mysterymath Can you please describe what this broke in the Fuchsia toolchain CI? Was it the data formatters?

**Comment by ldionne** [comment]
@Michael137 IIUC we'd be fine to re-submit now, is that right?

**Comment by Michael137** [comment]
> @Michael137 IIUC we'd be fine to re-submit now, is that right?

Not quite. I have a plan but havent put up a PR yet. Hopefully we're good by EOD tomorrow. I'll ping once I landed the fix

**Comment by Michael137** [comment]
PR is up: https://github.com/llvm/llvm-project/pull/108343

**Comment by Michael137** [comment]
Just merged the LLDB patch. Feel free to reland (this should also now be tested in the libc++ pre-merge CI)

**Pull request merged**
